fix for ow2-proactive/scheduling#2687

fviale opened the pull request:
A race condition can occur when shutting down the JVM. In that case, the VFSMountManagerHelper shutdown hook can execute concurrently with VFSMountManagerHelper.closeFileSystems. In that case, the vfsManager(VFS:DefaultFileSystemManager) can be closed by the shutdown hook and used later in the closeFileSystems method

This fix is meant to avoid such conflicting states by the following:
- shutdown hook now calls the terminate method which is synchronized.
- this terminate method nullifies the vfsManager field.
- the closeFileSystems method does nothing if the VFSMountManagerHelper is already terminated.
- The VFSMountManagerHelper can still be reused if a new call to mountAny is done, which regenerates the helper.

Added a test to verify that, after the terminate call, the closeFileSystems does not throw a NPE, and the helper can be regenerated with a call to mountAny

laurentpellegrino reviewed (changes requested):
This pull-request is not acceptable from my point of view.

First it includes a commit (entitled _Add reuseAddress option to prevent "Address already in use" errors wh_) that is really not related to the original purpose of the pull-request. This commit should be moved in another pull-request. Regarding its content, I am afraid to read that the new options which are introduced seem to be system dependent:
- http://docs.oracle.com/javase/7/docs/api/java/net/StandardSocketOptions.html?is-external=true#SO_REUSEADDR  
- http://stackoverflow.com/questions/14388706/socket-options-so-reuseaddr-and-so-reuseport-how-do-they-differ-do-they-mean-t/14388707#14388707

Second, the commit related to the original purpose of the pull-request contains no test. Having no test and a commit with a title that starts with _Tentative fix for..._ gives me the feeling that random programming was applied, even if I know that it is not the case. There should be a test to reproduce the issue, otherwise it is impossible to know if the patch fixes the issue or hide it.

fviale committed:
fix for ow2-proactive/scheduling#2687

A race condition can occur when shutting down the JVM. In that case, the VFSMountManagerHelper shutdown hook can execute concurrently with VFSMountManagerHelper.closeFileSystems. In that case, the vfsManager(VFS:DefaultFileSystemManager) can be closed by the shutdown hook and used later in the closeFileSystems method

This fix is meant to avoid such conflicting states by the following:
- shutdown hook now calls the terminate method which is synchronized.
- this terminate method nullifies the vfsManager field.
- the closeFileSystems method does nothing if the VFSMountManagerHelper is already terminated.
- The VFSMountManagerHelper can still be reused if a new call to mountAny is done, which regenerates the helper.

Added a test to verify that, after the terminate call, the closeFileSystems does not throw a NPE, and the helper can be regenerated with a call to mountAny
